app/data_sources/mattermost.py

```diff
@@ -0,0 +1,163 @@
+import logging
+from datetime import datetime
+from urllib.parse import urlparse
+from dataclasses import dataclass, asdict
+from typing import Dict, List, Optional
+
+from mattermostdriver import Driver
+
+from data_source_api.base_data_source import BaseDataSource, ConfigField, HTMLInputType
+from data_source_api.basic_document import BasicDocument, DocumentType
+from data_source_api.exception import InvalidDataSourceConfig
+from data_source_api.utils import parse_with_workers
+from indexing_queue import IndexingQueue
+
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class MattermostChannel:
+    id: str
+    name: str
+    team_id: str
+
+@dataclass
+class MattermostConfig:
+    url: str
+    token: str
+    scheme: Optional[str] = "https"
+    port: Optional[int] = 443
+    
+    def __post_init__(self):
+        try:
+            parsed_url = urlparse(self.url)
+        except Exception as e:
+            raise ValueError from e
+        
+        self.url = parsed_url.hostname
+        self.port = parsed_url.port if parsed_url.port is not None else self.port
+        self.scheme = parsed_url.scheme if parsed_url.scheme != "" else self.scheme
+
+        
+
+class MattermostDataSource(BaseDataSource):
+    FEED_BATCH_SIZE = 500
+
+    @staticmethod
+    def get_config_fields() -> List[ConfigField]:
+        return [
+            ConfigField(label="Mattermost Server", name="url", placeholder="https://mattermost.server.com", input_type=HTMLInputType.TEXT),
+            ConfigField(label="Token", name="token", placeholder="paste-your-token-here", input_type=HTMLInputType.PASSWORD),
+        ]
+
+
+    @staticmethod
+    def validate_config(config: Dict) -> None:
+        try:
+            parsed_config = MattermostConfig(**config)
+            maattermost = Driver(options = asdict(parsed_config))
+            maattermost.login()
+        except Exception as e:
+            raise InvalidDataSourceConfig from e
+
+    
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        mattermost_config = MattermostConfig(**self._config)
+        self._mattermost = Driver(options = asdict(mattermost_config))
+    
+    
+    def _list_channels(self) -> List[MattermostChannel]:
+        channels = self._mattermost.channels.client.get("/channels")
```

Comment:
do we need pagination here?

Reply:
Defaults to getting all channels, not sure if it's the smartest move